### PR TITLE
Apply package to ContributionProvider, FileResolver, PackageResolver, TemplateEngine

### PR DESCRIPTION
**Pull request by 928PJY**
#6835 

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/6841)

### CODE REVIEW

File: src/docfx/restore/FileResolver.cs [hunk 1]
```diff
@@ -178,14 +182,14 @@ private string DownloadFromUrlCore(string url)
             }
         }
 
-        private static string GetRestorePathFromUrl(string url)
+        private static PathString GetRestorePathFromUrl(string url)
         {
-            return PathUtility.NormalizeFile(Path.Combine(AppData.GetFileDownloadPath(url)));
+            return new PathString(PathUtility.NormalizeFile(Path.Combine(AppData.GetFileDownloadPath(url))));
```
Comment (yufeih) [inline src/docfx/restore/FileResolver.cs:187]: `new PathString` does the path normalization.
```suggestion
            return new PathString(Path.Combine(AppData.GetFileDownloadPath(url))));
```

**Review by yufeih — CHANGES_REQUESTED** [review]
Consider leave FileResolver and PackageResolver as is for now, they are typically external resources and so let's focus on content first.

**Comment by 928PJY** [comment]
Resouce used in FileResolver can be content as well.
As for packageResolver, template can also be a part of the repository, we should also handle it for now.

---
In reply to: [544666888](https://github.com/dotnet/docfx/pull/6841#issuecomment-544666888) [](ancestors = 544666888)

File: src/docfx/build/context/Input.cs [hunk 2]
```diff
@@ -107,6 +107,18 @@ public bool Exists(FilePath file)
...
+        public DateTime GetLastWriteTimeUtc(FilePath file)
+        {
+            if (file.IsGitCommit || file.Origin == FileOrigin.Generated)
+            {
+                return default;
+            }
+
+            var (package, path) = ResolveFilePath(_sourceMap.GetOriginalFilePath(file) ?? file);
+
+            return package.GetLastWriteTimeUtc(path);
```
Comment (yufeih) [inline src/docfx/build/context/Input.cs:119]:
```suggestion
            return package.TryGetLastWriteTimeUtc(path);
```

File: src/docfx/template/TemplateEngine.cs [hunk 3]
```diff
@@ -41,7 +41,11 @@ internal class TemplateEngine
             };
 
         public TemplateEngine(
```
Comment (yufeih) [inline src/docfx/template/TemplateEngine.cs:43]: no change?

**Review by yufeih — APPROVED** [review]

**Pull request merged**